### PR TITLE
svZeroDGUI Tool

### LINKED ISSUE #121: svZeroDGUI for 0d Model Generation
### Use Case

The svZeroDGUI web application allows users to generate input files for the svZeroDSolver by drawing the network on an easy-to-use GUI. This provides an alternative to manually creating files and is useful for users without access to a 3D model.

### Problem

Currently, there are two primary methods for creating 0D input files in SimVascular. The first method involves extracting the 0D model from an existing 3D model. This approach has significant drawbacks, including the requirement for a 3D model, which may not always be available. Additionally, this method is computationally expensive and time-intensive, which can be a barrier for many users.

The second method is manually creating the 0D input files, which is needed when a 3D model is not available or when a stand-alone simulation is desired. The lack of automation means that repetitive tasks must be done manually, which is inefficient and prone to errors. Furthermore, manually created files are difficult to debug and this approach lacks scalability, making it cumbersome to handle larger or more complex networks.

Given that 0D simulations are designed to be fast and efficient, these current methods do not fully align with these goals. An improved solution is needed to streamline the creation of 0D models, enhancing speed, efficiency, and scalability.

### Solution

The svZeroDGUI application addresses these needs by providing an intuitive graphical user interface for drawing the network and generating the input files. This solution:
- **Eliminates the Need for a 3D Model:** Allows users to create models without relying on an existing 3D model.
- **Automates File Generation:** Simplifies and speeds up the process of creating input files.
- **Enhances Scalability:** Facilitates the creation of larger and more complex networks efficiently.

The web application's frontend is built with HTML, CSS, and JavaScript, utilizing Cytoscape.js for network visualization and manipulation. The frontend features an intuitive node-linking system, where users can click on a node and drag an edge to connect it, making network construction straightforward. It also includes interactive parameter settings, enabling users to input vessel information, select junction types, and configure various network parameters directly within the application.

On the backend, Node.js handles server-side logic, while Cypress is employed for automated end-to-end testing to ensure the application functions correctly. The application incorporates an integrated validation framework that automatically checks the network for errors, such as missing elements or incorrect configurations, and provides error messages to assist users in correcting these issues. This setup ensures a user-friendly experience for building and editing 0D models while maintaining robust functionality and reliability.

https://github.com/user-attachments/assets/94dba682-8434-4d11-9932-576576279c2d

### Alternatives considered

We considered an alternative method for generating input files by training a computer vision model to parse hand-drawn circuit diagrams. Although this approach was effective for smaller inputs, we found that a GUI is more intuitive and user-friendly. The GUI not only simplifies the process of creating and editing the network structure but also facilitates the input of parameters directly into the files, rather than just generating the JSON structure. This makes it easier for users to manage and configure their models, improving the overall user experience.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct and Contributing Guidelines

### PR DESCRIPTION
**Pull request by emilinmathew**
Adding svZeroDGUI Tool

## Current situation
Addresses #121. Allows users to create 0D model input files through an intuitive graphical user interface.

## Release Notes 
- Java-based Web App: Utilizes Flask for backend, Node.js for server-side logic, and Cypress for testing.
- Node Management: Users can add nodes interactively within the graph representation.
- Edge Drawing: Intuitive UI for drawing edges or connections between nodes.
- Integrated Validation: Built-in validation framework to check consistency for svZeroDSolver.

## Documentation
Linked in docs/pages/GUI.md

## Testing
Passes all existing tests & added more unit tests written through Cypress for E2E testing web platforms. 

## Code of Conduct & Contributing Guidelines 
- [Yes] I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: .github/workflows/img2text.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# This workflow uses Cypress for end to end testing of the 0d model GUI.
+# Created by Emilin Mathew (emilinm@stanford.edu)
+
+name: e2e-tests
```
Comment (menon-karthik) [inline .github/workflows/img2text.yml]: Change to GUI-tests

File: .github/workflows/img2text.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+on: [push]
+jobs:
+  cypress-run:
+    runs-on: ubuntu-latest
```
Comment (menon-karthik) [inline .github/workflows/img2text.yml]: Run this on macos too? I can't remember if the other PR was also being tested only on Ubuntu, but if it is, maybe try to change that to ubuntu and macos?

File: .github/workflows/img2text.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      - name: Install dependencies
+        working-directory: tests/cypress
+        run: npm install
+
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+            python-version: '3.10'  # Use the Python version compatible with your Flask app
+
+      - name: Print Current Directory
```
Comment (menon-karthik) [inline .github/workflows/img2text.yml]: Is this required?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -35,23 +35,23 @@ jobs:
...
-          ./svZeroD_interface_test02 ../../../../Release ../../test_02/svzerod_tuned.json
+            cd tests/test_interface
+            mkdir build_tests
+            cd build_tests
+            cmake ../
+            make -j2
+            cd test_01
+            ./svZeroD_interface_test01 ../../../../Release ../../test_01/svzerod_3Dcoupling.json
+            cd ../test_02
+            ./svZeroD_interface_test02 ../../../../Release ../../test_02/svzerod_tuned.json
```
Comment (menon-karthik) [inline .github/workflows/test.yml]: Make indentation consistent

File: applications/create_0dmodel/app.py
Comment (menon-karthik) [inline applications/create_0dmodel/app.py:1]: I would change the directory to `applications/svZeroDGUI` (to match the name of the application mentioned in the documentation). Same for dirgraph

File: applications/create_0dmodel/static/css/style.css
Comment (menon-karthik) [inline applications/create_0dmodel/static/css/style.css:1]: (1) Add a comment at the top saying what this file is for; (2) add comments above each section with a sentence saying what that section is. 
Most user/developers (including me) are not familiar with CSS

File: applications/create_0dmodel/static/css/vessel_icon.png
Comment (menon-karthik) [inline applications/create_0dmodel/static/css/vessel_icon.png:1]: Are there two vessel images? I remember seeing the one in vessel.png during our discussions. What is this second one for?

File: cypress/downloads/graph_data.json
Comment (menon-karthik) [inline cypress/downloads/graph_data.json:1]: Is this an output of the cypress testing or is it the reference json used to compare with in the test? If it is an output of the tests, does it need to be stored in the repository?

File: docs/pages/GUI.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+responsive and user-friendly interface. It utilizes Cytoscape.js, a popular package for creating
+interactive elements and graphical networks.
+
+*  Backend: Flask app, Node.js for server-side logic, and Cypress for testing.
+This architecture supports an intuitive user experience for 
+generating and managing 0D input files through a graphical interface.
+
+
+# How to Use
+1. Navigate to the `applications folder` and then to the `create_0dmodel` subdirectory.
```
Comment (menon-karthik) [inline docs/pages/GUI.md]: Navigate to the `applications` folder (typo in the position of `)

File: docs/pages/GUI.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+2. Launch the `app.py` file.
```
Comment (menon-karthik) [inline docs/pages/GUI.md:33]: Provide a code block for steps 1-2?

File: applications/create_0dmodel/static/js/app.js
Comment (menon-karthik) [inline applications/create_0dmodel/static/js/app.js:1]: Add more documentation/comments to this file?

File: applications/create_0dmodel/static/js/cytoscape-edgehandles.js
Comment (menon-karthik) [inline applications/create_0dmodel/static/js/cytoscape-edgehandles.js:1]: Same as above: Add more documentation/comments to this file?

**Review by menon-karthik — CHANGES_REQUESTED** [review]
Looks good! Mostly comments about documentation.

File: applications/create_0dmodel/static/css/vessel_icon.png
Comment (emilinmathew) [reply applications/create_0dmodel/static/css/vessel_icon.png:1]: Deleted the second one! Was the first iteration image

File: cypress/downloads/graph_data.json
Comment (emilinmathew) [reply cypress/downloads/graph_data.json:1]: Deleted this folder - thanks for catching it! We don't need it. I used it for storing the downloaded files during development.

File: applications/svZeroDGUI/static/js/app.js [hunk 7]
```diff
@@ -60,22 +76,22 @@ function updateBoundaryConditionTypeVisibility() {
...
                 let nodeId = `${nodeType}-${node_count[nodeType]}`;
                 node_count[nodeType] += 1;
 
                 let color;
                 if (nodeType === 'boundary_condition') {
+                    // Sets the icons and edge color for each node
                     switch (boundaryConditionType) {
                         case 'FLOW':
-                            color = '#FF00FF'; //hexadecimal
+                            color = '#FF00FF';
```
Comment (menon-karthik) [inline applications/svZeroDGUI/static/js/app.js]: Would be good to retain the comment saying what color the code corresponds to (for all the colors below)

**Review by menon-karthik — CHANGES_REQUESTED** [review]
Almost there!

**Review by menon-karthik — APPROVED** [review]

**Pull request merged**